### PR TITLE
[SPARK-34573][SQL] Avoid global locking in SQLConf object for sqlConfEntries map

### PR DESCRIPTION
**Pull request by sparkyengine**
### What changes were proposed in this pull request?
In the `SQLConf` object, the `sqlConfEntries` map is globally synchronized (it is a Java `Collections.synchronizedMap`): any operation, including a get, will need to acquire the lock. 

An example of this is calling the `DatatType.sameType` method. This will trigger a check on `SQLConf.get.caseSensitiveAnalysis`. So every time we compare two datatypes with sameType, we hit a lock.

To avoid having multiple tasks locking on this, a better approach would be to use a map that does not lock on read (like a `ConcurrentHashMap`). This map implementation does not lock on read, and on write it only locks the map partially. The only lock that happens is on write on the same map key.

### Why are the changes needed?
Multiple tasks performing any operation that directly or indirectly trigger a query to the `SQLConf.sqlConfEntries` map, will require acquiring a global lock on that map. Something as easy as calling `DataType.sameType(...)` would be locking on the global `sqlConfEntries` lock of the `Collections.synchronizedMap`. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
No functionality change. Existing unit tests run normally.

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
This change could get away the lock wait you mentioned in the jira? If so, could you describe more in the PR description?

**Comment by gabrielenizzoli** [comment]
> This change could get away the lock wait you mentioned in the jira? If so, could you describe more in the PR description?

Added more descriptive PR text.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -53,22 +53,22 @@ import org.apache.spark.util.Utils
 
 object SQLConf {
 
-  private[sql] val sqlConfEntries = java.util.Collections.synchronizedMap(
-    new java.util.HashMap[String, ConfigEntry[_]]())
+  private[sql] val sqlConfEntries =
+    new ConcurrentHashMap[String, ConfigEntry[_]]()
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:57]: Seems a reasonable fix cc: @srowen @cloud-fan @viirya

**Review by srowen — APPROVED** [review]
OK pending tests

**Review by viirya — COMMENTED** [review]
looks reasonable. pending test.

**Comment by gabrielenizzoli** [comment]
> Kubernetes integration test status failure
> URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40196/

@srowen what does this mean? in jenkins the failure is:
```
Setting status of 4ffa2576782e410cef4d5cb3d8865362679ff11c to FAILURE with url https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40196/ and message: 'Build finished. '
FileNotFoundException means that the credentials Jenkins is using is probably wrong. Or the user account does not have write access to the repo.
org.kohsuke.github.GHFileNotFoundException: https://api.github.com/repos/apache/spark/statuses/4ffa2576782e410cef4d5cb3d8865362679ff11c {"message":"Not Found","documentation_url":"https://docs.github.com/rest/reference/repos#create-a-commit-status"}
```

**Comment by maropu** [comment]
> @srowen what does this mean? in jenkins the failure is:

The failure is not related to this PR, so it's okay just to ignore it.

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by HyukjinKwon** [comment]
Yeah seems good

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**